lib/tpm2_capability.c

```diff
@@ -38,33 +39,96 @@
+            break;
+        case TPM2_CAP_PCRS:
+            size = sizeof(dummy.data.assignedPCR.pcrSelections[0]);
+            break;
+        case TPM2_CAP_TPM_PROPERTIES:
+            size = sizeof(dummy.data.tpmProperties.tpmProperty[0]);
+            break;
+        case TPM2_CAP_PCR_PROPERTIES:
+            size = sizeof(dummy.data.pcrProperties.pcrProperty[0]);
+            break;
+        case TPM2_CAP_ECC_CURVES:
+            size = sizeof(dummy.data.eccCurves.eccCurves[0]);
+            break;
+        case TPM2_CAP_VENDOR_PROPERTY:
+            size = sizeof(dummy.data.intelPttProperty.property[0]);
+            break;
+        default:
+            size = 0;
+            LOG_ERR("Unable to determine property size for capability:  %d\n", capability);
+            break;
+    }
+
+    return size;
+}
+
 bool tpm2_capability_get (TSS2_SYS_CONTEXT *sapi_ctx,
         TPM2_CAP capability,
         UINT32 property,
         UINT32 count,
         TPMS_CAPABILITY_DATA *capability_data) {
 
-    TPMI_YES_NO            more_data;
-
-    TSS2_RC rval = TSS2_RETRY_EXP(Tss2_Sys_GetCapability (sapi_ctx,
-                                NULL,
-                                capability,
-                                property,
-                                count,
-                                &more_data,
-                                capability_data,
-                                NULL));
-
-    if (rval != TPM2_RC_SUCCESS) {
-        LOG_ERR("Failed to GetCapability: capability: 0x%x, property: 0x%x",
-                 capability, property);
-        LOG_PERR(Tss2_Sys_GetCapability, rval);
-        return false;
-    } else if (more_data) {
-        LOG_WARN("More data to be queried: capability: 0x%x, property: "
-                 "0x%x\n", capability, property);
-        return false;
-    }
+    TPMI_YES_NO more_data;
+
+    TPMS_CAPABILITY_DATA local_data;
+
+    UINT32 result_count;
+    UINT32 offset = 0;
+    size_t property_size = tpm2_get_property_data_size(capability);
+
+    /* All count and data fields line up within the TPMU_CAPABILITIES union
+        so it it safe to pick one here. */
+    void *data_dest = (void *) capability_data->data.algorithms.algProperties;
+    void *data_src  = (void *) local_data.data.algorithms.algProperties;
+
+    do {
+        TSS2_RC rval = TSS2_RETRY_EXP(Tss2_Sys_GetCapability (sapi_ctx,
+                    NULL,
+                    capability,
+                    property+offset,
+                    count,
+                    &more_data,
+                    &local_data,
+                    NULL));
+
+        if (rval != TPM2_RC_SUCCESS) {
+            LOG_ERR("Failed to GetCapability: capability: 0x%x, property: 0x%x",
+                  capability, property);
+            LOG_PERR(Tss2_Sys_GetCapability, rval);
+            return false;
+        }
+
+        result_count = local_data.data.algorithms.count;
+
+        data_dest += offset * property_size;
```

Comment:
The fact that it's letting you do pointer arithmetic on a `void *` blows my mind. I think you want `data_dest` and `data_src` to be of type `BYTE *` Unless you can point to the C99 spec that says this arithmetic is defined in the standard as one memory address. Normal pointer arithmetic moves the address ahead by sizeof the dereferenced type, which in void doesn't have one.

Reply:
I think at least the GCC compiler accepts this, but I agree it is not that nice to do pointer arithmetic on a void *.  So I will make this more clear as well by modifying the pointer type.